cpp/src/arrow/compute/exec/exec_plan.cc

```diff
@@ -283,6 +290,35 @@ bool ExecNode::ErrorIfNotOk(Status status) {
   return true;
 }
 
+Status ExecNode::SubmitTask(std::function<Status()> task) {
+  if (finished_.is_finished()) {
+    return Status::OK();
+  }
+  if (this->has_executor()) {
+    DCHECK(task_group_ != nullptr);
+    task_group_->Append(std::move(task));
+  } else {
+    std::move(task)();
+  }
+  if (batch_count_.Increment()) {
+    this->MarkFinished();
+  }
+  return Status::OK();
+}
+
+void ExecNode::MarkFinished(bool request_stop) {
+  if (this->has_executor()) {
+    if (request_stop) {
+      this->stop_source_.RequestStop();
+    }
+    task_group_->FinishAsync().AddCallback([this](const Status& status) {
+      if (!this->finished_.is_finished()) this->finished_.MarkFinished(status);
```

Comment:
This is racy/this is a TOC-TOU error.

Reply:
So long as this is only called after checking AtomicCounter, is the check here necessary?

Reply:
Based on my test and profiling test, yes this is necessary. The first call to this functions could when `batch_count_` reach the total_batches, but at the same time ExecNode::StopProducing signal could happens. So only once it's neccesary to mark as Finished  `finished_`.

Reply:
Hmm. I think we have a race, then. The point of AtomicCounter is so that we only ever make one call to MarkFinished. I also just noticed that batch_count_.Increment() needs to go inside the task, not outside it.